# FareCast NYC: Identifying Generous Tippers with Predictive Modeling

## Overview

**Executive Summary:**

In this machine learning endeavor, our primary objective is to empower taxi cab drivers by developing a predictive model that identifies and predicts generous tipping behavior among passengers. The ultimate goal is to enhance the revenue generated by taxi drivers, particularly through increased earnings from tips. To achieve this, we aim to construct a model capable of discerning customers who exhibit an exceptionally generous tipping inclination—specifically those who tip 20% or more of the fare amount. By leveraging machine learning techniques, we anticipate uncovering patterns and factors that contribute to a passenger's generosity, providing valuable insights to drivers and taxi service providers. This model not only supports individual drivers in customizing their service but also contributes to the broader optimization of the taxi service industry by fostering a more rewarding experience for both drivers and passengers.

**Ethical Considerations**

In lieu of tip estimation, the focus shifts to identifying generous customers for ethical considerations. Predicting tips poses potential challenges, as drivers may be disheartened if the app incorrectly informs them about a customer's intent to tip. This could erode trust in the application, impacting the overall relationship between drivers and the service. Moreover, if drivers begin avoiding passengers predicted not to tip, it may lead to accessibility issues for customers, especially those who cannot afford additional tips. The model's predictions could inadvertently influence driver behavior, affecting the equitable provision of taxi services. Choosing to identify generous customers mitigates these potential issues, aligning with ethical considerations to ensure fair and unbiased service delivery.

**Data Sources:**

This project relies on the "2017 Yellow Taxi Trip Data" as the primary dataset, sourced from NYC OpenData. The dataset, processed and exported from `fare_estimation.ipynb`, provides comprehensive information about taxi and for-hire vehicle trips.

For a detailed understanding of each attribute and its significance, please refer to the [Data Dictionary](https://github.com/sssunri/farecast_NYC-TLC_estimation/blob/main/data_dictionary_trip_records_yellow.pdf) presented by the New York City Taxi & Limousine Commission. The richness of this data dictionary is instrumental in comprehending the dataset, a crucial aspect in constructing a reliable fare estimation model.

To facilitate accurate prediction, the model requires historical behavioral data for each customer, including tipping patterns from previous taxi rides. Essential features encompass timestamps, dates, locations of pickups and dropoffs, estimated fares, and payment methods. The binary target variable signifies whether the customer is anticipated to tip ≥ 20%.

## Data Exploration

**Load the Data**

In [1]:
# packages for numerics + dataframes
import pandas as pd
import numpy as np
from scipy import stats

pd.set_option('display.max_columns', None)

# packages for visualization
import matplotlib.pyplot as plt
import seaborn as sns

plt.rcParams['figure.figsize'] = 15, 5

# packages for OLS, MLR, confusion matrix
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.metrics import roc_auc_score, roc_curve, RocCurveDisplay
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay

# function that helps plot feature importance 
from xgboost import plot_importance

# save your models and read them back in
from pathlib import Path
import pickle

In [2]:
# load dataset into dataframe
df = pd.read_csv('data/2017_Yellow_Taxi_Trip_Data_Export.csv')

**Initial Exploration**

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22644 entries, 0 to 22643
Data columns (total 29 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   VendorID               22644 non-null  int64  
 1   tpep_pickup_datetime   22644 non-null  object 
 2   tpep_dropoff_datetime  22644 non-null  object 
 3   passenger_count        22644 non-null  int64  
 4   trip_distance          22644 non-null  float64
 5   RatecodeID             22644 non-null  int64  
 6   store_and_fwd_flag     22644 non-null  object 
 7   PULocationID           22644 non-null  int64  
 8   DOLocationID           22644 non-null  int64  
 9   payment_type           22644 non-null  int64  
 10  fare_amount            22644 non-null  float64
 11  extra                  22644 non-null  float64
 12  tip_amount             22644 non-null  float64
 13  tolls_amount           22644 non-null  float64
 14  total_amount           22644 non-null  float64
 15  mo

## Feature Engineering

**Analysis of Average Tips Based on Payment Type**

In [4]:
# what is the average tip for trips paid for with credit card?
credit_avg_tip = df[df['payment_type'] == 1].mean(numeric_only=True)['tip_amount']
print(f'credit card average tip: ${credit_avg_tip:.2f}')

# what is the average tip for trips paid for with cash?
cash_avg_tip = df[df['payment_type'] == 2].mean(numeric_only=True)['tip_amount']
print(f'cash average tip: ${cash_avg_tip}')

credit card average tip: $2.70
cash average tip: $0.0


Examining the tipping patterns based on payment methods reveals interesting insights. On average, trips paid with credit cards receive a substantial tip of \\$2.70. In contrast, cash payments show an apparent difference, with an average tip of \\$0.00. This stark contrast in tipping behavior between credit card and cash transactions highlights a potential distinction in customer preferences and tipping practices.

To streamline the modeling objective of predicting generous tippers, we will narrow down our focus by sampling data exclusively from customers who opt for credit card payments. This targeted approach aims to capture patterns specific to credit card transactions, providing a more refined dataset for accurate prediction and analysis.

In [5]:
# subset the data to isolate only customers who paid by credit card
df = df[df['payment_type'] == 1].copy()
df['payment_type'].unique()

array([1], dtype=int64)

**Creating Binary Indicator for Generous Tips**

To better facilitate the modeling objective, we will introduce a new binary column, `generous_tip`, serving as the target variable. This column will act as a clear indicator, assigning a value of 1 if the customer tipped 20% or more and 0 otherwise. By incorporating this binary classification, the model can precisely identify and predict customers who exhibit generous tipping behavior, streamlining the focus on those likely to leave substantial tips.

In [6]:
df['generous_tip'] = df['tip_percent'].copy()
df['generous_tip'] = (df['tip_percent'] >= 0.2).astype(int)
df['generous_tip'].head()

0    1
1    1
2    0
3    1
5    1
Name: generous_tip, dtype: int32

In [7]:
df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2017-03-25 08:55:43   2017-03-25 09:09:47                6   
1         1  2017-04-11 14:53:28   2017-04-11 15:19:58                1   
2         1  2017-12-15 07:26:56   2017-12-15 07:34:08                1   
3         2  2017-05-07 13:17:59   2017-05-07 13:48:14                1   
5         2  2017-03-25 20:34:11   2017-03-25 20:42:11                6   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           3.34           1                  N           100           231   
1           1.80           1                  N           186            43   
2           1.00           1                  N           262           236   
3           3.70           1                  N           188            97   
5           2.30           1                  N           161           236   

   payment_type  fare_amount  extra  tip_amount  tolls_amount  total_amount  \
0             1         13.0    0.0        2.76           0.0         16.56   
1             1         16.0    0.0        4.00           0.0         20.80   
2             1          6.5    0.0        1.45           0.0          8.75   
3             1         20.5    0.0        6.39           0.0         27.69   
5             1          9.0    0.5        2.06           0.0         12.36   

      month       day  hour time_of_day  season  duration pickup_dropoff  \
0     March  Saturday     8     am_rush       2     14.07        100 231   
1     April   Tuesday    14     daytime       2     26.50         186 43   
2  December    Friday     7     am_rush       1      7.20        262 236   
3       May    Sunday    13     daytime       2     30.25         188 97   
5     March  Saturday    20   nighttime       2      8.00        161 236   

   mean_distance  mean_duration  tip_percent  trip_speed_mph        date  \
0       3.521667      22.848333        0.200       14.243070  2017-03-25   
1       3.108889      24.470000        0.238        4.075472  2017-04-11   
2       0.881429       7.250286        0.199        8.333333  2017-12-15   
3       3.700000      30.250000        0.300        7.338843  2017-05-07   
5       2.052258      11.855484        0.200       17.250000  2017-03-25   

   rush_hour  predicted_fare  generous_tip  
0          0       13.636419             1  
1          0       13.916246             1  
2          1        7.128055             0  
3          0       18.647323             1  
5          0        9.921377             1

**Acknowledgments**

This project is inspired by the [Google Advanced Data Analytics Professional Certificate](https://www.coursera.org/professional-certificates/google-advanced-data-analytics) program, specifically the end-of-course portfolio project titled *Automatidata*.

This project utilises a subset of the [2017 Yellow Taxi Trip Data](https://data.cityofnewyork.us/Transportation/2017-Yellow-Taxi-Trip-Data/biws-g3hs) obtained from the New York City Taxi & Limousine Commission, originally published as part of the NYC Open Data program. The dataset, containing 22,699 rows representing different trips and 18 columns, was provided by the Google Advanced Data Analytics Professional Certificate program for the purpose of the end-of-course portfolio project. Invaluable insights into the dataset's attributes and their meanings were derived from the detailed [Data Dictionary](https://data.cityofnewyork.us/api/views/biws-g3hs/files/eb3ccc47-317f-4b2a-8f49-5a684b0b1ecc?download=true&filename=data_dictionary_trip_records_yellow.pdf) provided by the New York City Taxi & Limousine Commission. We acknowledge the contribution of both the NYC TLC and the Google program to the dataset used in this analysis.